### PR TITLE
[Fit API] improve event handlers

### PR DESCRIPTION
**Pull request by roywei**
## Description ##
Making the follwing on evetn handlers based on the design here:
https://cwiki.apache.org/confluence/display/MXNET/Callback+Design+for+Fit+Loop

1. Making metric update and validation logic in event handlers
2. Each event handler maintain it's own states
3. Passing a weak reference of estimator at each callback call, so some attributes are passed(net, trainer, etc)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/gluon/contrib/estimator/estimator.py [hunk 1]
```diff
@@ -57,87 +60,85 @@ def __init__(self, net,
                  context=None):
 
         self.net = net
+        self.loss = self._check_loss(loss)
+        self.train_metrics = self._check_metrics(metrics)
+
+        # Use default mx.metric.Accuracy() for gluon.loss.SoftmaxCrossEntropyLoss()
+        if not self.train_metrics and any([isinstance(l, gluon.loss.SoftmaxCrossEntropyLoss) for l in self.loss]):
```
Comment (szha) [inline python/mxnet/gluon/contrib/estimator/estimator.py]: instead of hard coding it here, how about specify it as a default argument?

Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/estimator.py]: @szha  Agree, but it Acc as default may not work well for other type of loss. How about moving this logic to prepare_loss_and_metrics? So the reference is returned to user.

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 2]
```diff
@@ -16,85 +16,169 @@
...
 
     LOG_VERBOSITY_PER_EPOCH = 1
     LOG_VERBOSITY_PER_BATCH = 2
 
-    def __init__(self, file_name=None, file_location=None, verbose=LOG_VERBOSITY_PER_EPOCH):
+    def __init__(self, file_name=None,
+                 file_location=None,
+                 verbose=LOG_VERBOSITY_PER_EPOCH,
+                 train_metrics=None,
+                 val_metrics=None):
```
Comment (nswamy) [inline python/mxnet/gluon/contrib/estimator/event_handler.py]: is it possible to customize, given that prepare_loss_and_metrics happen after the estimator is created. If at all the order should be
```
e = Estimator()
e.prepare_loss_and_metrics()
lh = LoggingHandler(..., train_metrics=[e.train_metrics[0], e.train_metrics[1]], ..)
```
is this what you were thinking?

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 3]
```diff
@@ -112,71 +196,83 @@ def __init__(self, file_name=None, file_location=None, verbose=LOG_VERBOSITY_PER
...
             msg += 'time/batch: %.3fs ' % batch_time
-            for key in self.estimator.train_stats:
+            for metric in self.train_metrics:
                 # only log current training loss & metric after each batch
-                if key.startswith('train_'):
-                    msg += key + ': ' + '%.4f ' % self.estimator.train_stats[key]
+                name, value = metric.get()
+                msg += '%s : %.4f ' % (name, value)
             self.logger.info(msg)
+            self.batch_index += 1
```
Comment (nswamy) [inline python/mxnet/gluon/contrib/estimator/event_handler.py]: shouldn't this be in the estimator itself, why should all handlers maintain this?

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 4]
```diff
@@ -219,55 +320,61 @@ def __init__(self,
...
             if self.save_best_only:
+                monitor_name, monitor_value = self.monitor.get()
                 # check if monitor exists in train stats
-                if self.monitor not in self.estimator.train_stats:
-                    warnings.warn(RuntimeWarning('Unable to find %s in training statistics, make sure the monitor value'
-                                                 'starts with `train_ `or `val_` and contains loss/metric name, ',
-                                                 'for example val_accuracy', self.monitor))
-                    self.estimator.net.save_parameters(self.filepath)
+                if np.isnan(monitor_value):
+                    warnings.warn(RuntimeWarning('%s is not updated, make sure you pass one of the metric objects'
```
Comment (nswamy) [inline python/mxnet/gluon/contrib/estimator/event_handler.py]: use logger, so the user can control.

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 5]
```diff
@@ -191,18 +287,23 @@ class CheckpointHandler(EventHandler):
 
     def __init__(self,
                  filepath,
-                 monitor='val_accuracy',
+                 monitor=None,
                  verbose=0,
                  save_best_only=False,
                  mode='auto',
```
Comment (nswamy) [inline python/mxnet/gluon/contrib/estimator/event_handler.py]: can you expand what different modes mean in the doc.

File: python/mxnet/gluon/contrib/estimator/estimator.py [hunk 6]
```diff
@@ -148,63 +145,70 @@ def _is_initialized(self):
...
+            if not isinstance(val_data, gluon.data.DataLoader):
+                raise ValueError("Estimator only support input as Gluon DataLoader. Alternatively, you "
+                                 "can transform your DataIter or any NDArray into Gluon DataLoader. "
+                                 "Refer to gluon.data.dataloader")
+            data, label = self._get_data_and_label(batch, self.context)
             pred = [self.net(x) for x in data]
-            losses = []
-            for loss in self.loss:
-                losses.append([loss(y_hat, y) for y_hat, y in zip(pred, label)])
+            loss = [self.loss[0](y_hat, y) for y_hat, y in zip(pred, label)]
```
[the remark this answers is not in the document]
Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/estimator.py]: multi loss will be supported in https://github.com/apache/incubator-mxnet/pull/14628, let's get the first version into master and iterate on that.

File: python/mxnet/gluon/contrib/estimator/estimator.py [hunk 7]
```diff
@@ -226,111 +230,72 @@ def fit(self, train_data,
...
-                          "Please look at gluon.estimator.event_handler for more detail.")
+        if not event_handlers:
+            train_metrics, val_metrics = self.prepare_loss_and_metrics()
+            event_handlers.append(MetricHandler(train_metrics=train_metrics))
+            if val_data:
+                event_handlers.append(ValidationHandler(val_data=val_data, eval_fn=self.evaluate,
+                                                        val_metrics=val_metrics))
+            event_handlers.append(LoggingHandler(train_metrics=train_metrics,
+                                                 val_metrics=val_metrics))
+            warnings.warn("No Event Handler specified, default %s are used. "
```
[the remark this answers is not in the document]
Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/estimator.py]: Good point! for now we can only do this for estimator and handlers, any other warning from mxnet and gluon still  can't be controlled. tracked here: https://issues.apache.org/jira/browse/MXNET-1395

File: python/mxnet/gluon/contrib/estimator/estimator.py [hunk 8]
```diff
@@ -226,111 +230,72 @@ def fit(self, train_data,
...
                 for handler in batch_begin:
-                    handler.batch_begin()
+                    handler.batch_begin(estimator_ref, batch=batch)
 
                 with autograd.record():
                     pred = [self.net(x) for x in data]
-                    losses = []
-                    for loss in self.loss:
-                        losses.append([loss(y_hat, y) for y_hat, y in zip(pred, label)])
+                    loss = [self.loss[0](y_hat, y) for y_hat, y in zip(pred, label)]
```
[the remark this answers is not in the document]
Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/estimator.py]: as above https://issues.apache.org/jira/browse/MXNET-1395

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 9]
```diff
@@ -16,85 +16,169 @@
...
+        self.train_metrics = train_metrics or []
+        # order to be called among all callbacks
+        # metrics need to be calculated before other callbacks can access them
+        self.priority = -np.Inf
+
+    def epoch_begin(self, estimator, *args, **kwargs):
+        for metric in self.train_metrics:
+            metric.reset()
+
+    def batch_end(self, estimator, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/event_handler.py]: once batch end we lost that batch's label and pred

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 10]
```diff
@@ -16,85 +16,169 @@
...
+    ----------
+    train_metrics : List of EvalMetrics
+        training metrics to be updated at batch end
+    """
+
+    def __init__(self, train_metrics):
+        self.train_metrics = train_metrics or []
+        # order to be called among all callbacks
+        # metrics need to be calculated before other callbacks can access them
+        self.priority = -np.Inf
```
[the remark this answers is not in the document]
Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/event_handler.py]: This mechanism is mainly for internal use. I m making sure metric and validation are called first and logging are called last. I'm trying to reduce the information user need to know, he can order his own event handlers in the list  before passing to fit()

File: python/mxnet/gluon/contrib/estimator/event_handler.py [hunk 11]
```diff
@@ -16,85 +16,169 @@
...
+                 eval_fn,
+                 val_metrics=None,
+                 epoch_period=1,
+                 batch_period=None):
+        self.val_data = val_data
+        self.eval_fn = eval_fn
+        self.epoch_period = epoch_period
+        self.batch_period = batch_period
+        self.val_metrics = val_metrics
+        self.num_batches = 0
```
[the remark this answers is not in the document]
Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/event_handler.py]: good point! if we provide default handler one by one (so user don't need to re-create all just to custom one of them). We need a mechanism to make sure all handlers has the reference of the same set of metric objects.  or make handlers an attributes so they can be configured after default handlers been created. tracked https://issues.apache.org/jira/browse/MXNET-1396

File: python/mxnet/gluon/contrib/estimator/event_handler.py (hunk 2 above)
Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/event_handler.py]: yes that's correct

File: python/mxnet/gluon/contrib/estimator/event_handler.py (hunk 3 above)
Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/event_handler.py]: not all handlers need the same set of these infomation, so they maintain whatever they want to use. This also prevents if one handler changed `self.estimator.total_steps` wrongly, it will cause all other handlers to fail

File: python/mxnet/gluon/contrib/estimator/event_handler.py (hunk 5 above)
Comment (roywei) [reply python/mxnet/gluon/contrib/estimator/event_handler.py]: will update, thanks!

File: python/mxnet/gluon/contrib/estimator/event_handler.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (nswamy) [reply python/mxnet/gluon/contrib/estimator/event_handler.py]: lets call this out explicitly in the documentation.
